{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Task 1. Nucleotide distribution\n",
    "\n",
    "nucl = pd.read_csv('/Users/Sasha/Downloads/train.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>reads_all</th>\n",
       "      <th>matches</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>deletions</th>\n",
       "      <th>insertions</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>A_fraction</th>\n",
       "      <th>T_fraction</th>\n",
       "      <th>G_fraction</th>\n",
       "      <th>C_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>279</td>\n",
       "      <td>8045</td>\n",
       "      <td>32</td>\n",
       "      <td>7972</td>\n",
       "      <td>46</td>\n",
       "      <td>8</td>\n",
       "      <td>7911.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.983503</td>\n",
       "      <td>0.000815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>280</td>\n",
       "      <td>8045</td>\n",
       "      <td>7995</td>\n",
       "      <td>48</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001355</td>\n",
       "      <td>0.002392</td>\n",
       "      <td>0.001914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>281</td>\n",
       "      <td>8045</td>\n",
       "      <td>7983</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000639</td>\n",
       "      <td>0.002875</td>\n",
       "      <td>0.002310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>282</td>\n",
       "      <td>8049</td>\n",
       "      <td>7510</td>\n",
       "      <td>505</td>\n",
       "      <td>33</td>\n",
       "      <td>65</td>\n",
       "      <td>105.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>395.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012750</td>\n",
       "      <td>0.049096</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>283</td>\n",
       "      <td>8042</td>\n",
       "      <td>7786</td>\n",
       "      <td>155</td>\n",
       "      <td>115</td>\n",
       "      <td>62</td>\n",
       "      <td>46.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.004960</td>\n",
       "      <td>0.010992</td>\n",
       "      <td>0.003186</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos  reads_all  matches  mismatches  deletions  insertions       A     C  \\\n",
       "0  279       8045       32        7972         46           8  7911.0  54.0   \n",
       "1  280       8045     7995          48         19           9     NaN  16.0   \n",
       "2  281       8045     7983          50         16           5     NaN  24.0   \n",
       "3  282       8049     7510         505         33          65   105.0  11.0   \n",
       "4  283       8042     7786         155        115          62    46.0   NaN   \n",
       "\n",
       "       T     G  A_fraction  T_fraction  G_fraction  C_fraction  \n",
       "0   11.0   NaN    0.983503    0.000815         NaN    0.006896  \n",
       "1   13.0  19.0         NaN    0.001355    0.002392    0.001914  \n",
       "2    4.0  24.0         NaN    0.000639    0.002875    0.002310  \n",
       "3  395.0   NaN    0.012750    0.049096         NaN    0.000919  \n",
       "4   93.0  31.0    0.004960    0.010992    0.003186         NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_nuc = pd.DataFrame()\n",
    "count_nuc['A'] = nucl['A']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>7911.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>NaN</td>\n",
       "      <td>24.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>105.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>395.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>46.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.0</td>\n",
       "      <td>93.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          A     C     G      T\n",
       "pos                           \n",
       "279  7911.0  54.0   NaN   11.0\n",
       "280     NaN  16.0  19.0   13.0\n",
       "281     NaN  24.0  24.0    4.0\n",
       "282   105.0  11.0   NaN  395.0\n",
       "283    46.0   NaN  31.0   93.0"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_nuc['C'] = nucl['C']\n",
    "count_nuc['G'] = nucl['G']\n",
    "count_nuc['T'] = nucl['T']\n",
    "count_nuc['pos'] = nucl['pos']\n",
    "count_nuc = count_nuc.set_index('pos')\n",
    "count_nuc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-5e130445ec7f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcolors\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'green'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'blue'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'lime'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m plt.hist(x, n_bins, density = True, \n\u001b[0m\u001b[1;32m      4\u001b[0m          \u001b[0mhisttype\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m'bar'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m          \u001b[0mcolor\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcolors\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "count_nuc.plot.bar(rot=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>reads_all</th>\n",
       "      <th>matches</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>deletions</th>\n",
       "      <th>insertions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>280</td>\n",
       "      <td>8045</td>\n",
       "      <td>7995</td>\n",
       "      <td>48</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>281</td>\n",
       "      <td>8045</td>\n",
       "      <td>7983</td>\n",
       "      <td>50</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>282</td>\n",
       "      <td>8049</td>\n",
       "      <td>7510</td>\n",
       "      <td>505</td>\n",
       "      <td>33</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>283</td>\n",
       "      <td>8042</td>\n",
       "      <td>7786</td>\n",
       "      <td>155</td>\n",
       "      <td>115</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>284</td>\n",
       "      <td>8048</td>\n",
       "      <td>7094</td>\n",
       "      <td>104</td>\n",
       "      <td>864</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos  reads_all  matches  mismatches  deletions  insertions\n",
       "1  280       8045     7995          48         19           9\n",
       "2  281       8045     7983          50         16           5\n",
       "3  282       8049     7510         505         33          65\n",
       "4  283       8042     7786         155        115          62\n",
       "5  284       8048     7094         104        864         241"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Task 2. Extract specific columns\n",
    "part_nuc = pd.DataFrame()\n",
    "part_nuc = nucl.loc[nucl['matches'] > np.mean(nucl['matches'])]\n",
    "part_nuc = part_nuc.drop(['A','T','G','C'], axis=1)\n",
    "part_nuc = part_nuc.drop(['A_fraction','T_fraction','G_fraction', 'C_fraction'],axis=1)\n",
    "part_nuc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 63 entries, 0 to 62\n",
      "Data columns (total 14 columns):\n",
      " #   Column      Non-Null Count  Dtype  \n",
      "---  ------      --------------  -----  \n",
      " 0   pos         63 non-null     int64  \n",
      " 1   reads_all   63 non-null     int64  \n",
      " 2   matches     63 non-null     int64  \n",
      " 3   mismatches  63 non-null     int64  \n",
      " 4   deletions   63 non-null     int64  \n",
      " 5   insertions  63 non-null     int64  \n",
      " 6   A           45 non-null     float64\n",
      " 7   C           48 non-null     float64\n",
      " 8   T           49 non-null     float64\n",
      " 9   G           47 non-null     float64\n",
      " 10  A_fraction  45 non-null     float64\n",
      " 11  T_fraction  49 non-null     float64\n",
      " 12  G_fraction  47 non-null     float64\n",
      " 13  C_fraction  48 non-null     float64\n",
      "dtypes: float64(8), int64(6)\n",
      "memory usage: 7.0 KB\n"
     ]
    }
   ],
   "source": [
    "# Task 3, EDA\n",
    "nucl.info()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>reads_all</th>\n",
       "      <th>matches</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>deletions</th>\n",
       "      <th>insertions</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>A_fraction</th>\n",
       "      <th>T_fraction</th>\n",
       "      <th>G_fraction</th>\n",
       "      <th>C_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>48.000000</td>\n",
       "      <td>49.000000</td>\n",
       "      <td>47.000000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>49.000000</td>\n",
       "      <td>47.000000</td>\n",
       "      <td>48.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>310.000000</td>\n",
       "      <td>8050.936508</td>\n",
       "      <td>4710.285714</td>\n",
       "      <td>1796.158730</td>\n",
       "      <td>1553.698413</td>\n",
       "      <td>109.666667</td>\n",
       "      <td>594.044444</td>\n",
       "      <td>380.666667</td>\n",
       "      <td>492.163265</td>\n",
       "      <td>949.212766</td>\n",
       "      <td>0.073241</td>\n",
       "      <td>0.060566</td>\n",
       "      <td>0.117431</td>\n",
       "      <td>0.046797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>18.330303</td>\n",
       "      <td>6.045187</td>\n",
       "      <td>2211.987374</td>\n",
       "      <td>1518.102245</td>\n",
       "      <td>1107.108288</td>\n",
       "      <td>149.519014</td>\n",
       "      <td>1295.866838</td>\n",
       "      <td>333.817719</td>\n",
       "      <td>582.504769</td>\n",
       "      <td>929.810942</td>\n",
       "      <td>0.161150</td>\n",
       "      <td>0.072425</td>\n",
       "      <td>0.115425</td>\n",
       "      <td>0.041395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>279.000000</td>\n",
       "      <td>8039.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>48.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>11.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>0.000639</td>\n",
       "      <td>0.002243</td>\n",
       "      <td>0.000919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>294.500000</td>\n",
       "      <td>8046.000000</td>\n",
       "      <td>3025.500000</td>\n",
       "      <td>656.500000</td>\n",
       "      <td>512.000000</td>\n",
       "      <td>21.500000</td>\n",
       "      <td>105.000000</td>\n",
       "      <td>154.250000</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>314.000000</td>\n",
       "      <td>0.012750</td>\n",
       "      <td>0.013547</td>\n",
       "      <td>0.038555</td>\n",
       "      <td>0.018568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>310.000000</td>\n",
       "      <td>8051.000000</td>\n",
       "      <td>4665.000000</td>\n",
       "      <td>1619.000000</td>\n",
       "      <td>1556.000000</td>\n",
       "      <td>53.000000</td>\n",
       "      <td>241.000000</td>\n",
       "      <td>278.500000</td>\n",
       "      <td>361.000000</td>\n",
       "      <td>663.000000</td>\n",
       "      <td>0.030059</td>\n",
       "      <td>0.044339</td>\n",
       "      <td>0.082288</td>\n",
       "      <td>0.033901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>325.500000</td>\n",
       "      <td>8054.500000</td>\n",
       "      <td>6737.500000</td>\n",
       "      <td>2453.500000</td>\n",
       "      <td>2300.000000</td>\n",
       "      <td>123.000000</td>\n",
       "      <td>480.000000</td>\n",
       "      <td>533.000000</td>\n",
       "      <td>634.000000</td>\n",
       "      <td>1185.000000</td>\n",
       "      <td>0.059539</td>\n",
       "      <td>0.078508</td>\n",
       "      <td>0.146423</td>\n",
       "      <td>0.065621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>341.000000</td>\n",
       "      <td>8062.000000</td>\n",
       "      <td>8002.000000</td>\n",
       "      <td>7972.000000</td>\n",
       "      <td>4215.000000</td>\n",
       "      <td>696.000000</td>\n",
       "      <td>7911.000000</td>\n",
       "      <td>1423.000000</td>\n",
       "      <td>3054.000000</td>\n",
       "      <td>4226.000000</td>\n",
       "      <td>0.983503</td>\n",
       "      <td>0.378969</td>\n",
       "      <td>0.524417</td>\n",
       "      <td>0.176147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              pos    reads_all      matches   mismatches    deletions  \\\n",
       "count   63.000000    63.000000    63.000000    63.000000    63.000000   \n",
       "mean   310.000000  8050.936508  4710.285714  1796.158730  1553.698413   \n",
       "std     18.330303     6.045187  2211.987374  1518.102245  1107.108288   \n",
       "min    279.000000  8039.000000    32.000000    48.000000     2.000000   \n",
       "25%    294.500000  8046.000000  3025.500000   656.500000   512.000000   \n",
       "50%    310.000000  8051.000000  4665.000000  1619.000000  1556.000000   \n",
       "75%    325.500000  8054.500000  6737.500000  2453.500000  2300.000000   \n",
       "max    341.000000  8062.000000  8002.000000  7972.000000  4215.000000   \n",
       "\n",
       "       insertions            A            C            T            G  \\\n",
       "count   63.000000    45.000000    48.000000    49.000000    47.000000   \n",
       "mean   109.666667   594.044444   380.666667   492.163265   949.212766   \n",
       "std    149.519014  1295.866838   333.817719   582.504769   929.810942   \n",
       "min      3.000000    14.000000    11.000000     4.000000    17.000000   \n",
       "25%     21.500000   105.000000   154.250000   118.000000   314.000000   \n",
       "50%     53.000000   241.000000   278.500000   361.000000   663.000000   \n",
       "75%    123.000000   480.000000   533.000000   634.000000  1185.000000   \n",
       "max    696.000000  7911.000000  1423.000000  3054.000000  4226.000000   \n",
       "\n",
       "       A_fraction  T_fraction  G_fraction  C_fraction  \n",
       "count   45.000000   49.000000   47.000000   48.000000  \n",
       "mean     0.073241    0.060566    0.117431    0.046797  \n",
       "std      0.161150    0.072425    0.115425    0.041395  \n",
       "min      0.001060    0.000639    0.002243    0.000919  \n",
       "25%      0.012750    0.013547    0.038555    0.018568  \n",
       "50%      0.030059    0.044339    0.082288    0.033901  \n",
       "75%      0.059539    0.078508    0.146423    0.065621  \n",
       "max      0.983503    0.378969    0.524417    0.176147  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-e508f45dbf18>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mncols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0maxs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflatten\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mnucl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mboxplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnucl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "fig, axs = plt.subplots(ncols=3, nrows=5, figsize=(20, 10))\n",
    "index = 0\n",
    "axs = axs.flatten()\n",
    "for i,j in nucl.items():\n",
    "    sns.boxplot(y=i, data=nucl, ax=axs[index])\n",
    "    index += 1\n",
    "plt.tight_layout(pad=0.4, w_pad=0.5, h_pad=5.0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A_fraction    18\n",
       "A             18\n",
       "G_fraction    16\n",
       "G             16\n",
       "C_fraction    15\n",
       "C             15\n",
       "T_fraction    14\n",
       "T             14\n",
       "insertions     0\n",
       "deletions      0\n",
       "mismatches     0\n",
       "matches        0\n",
       "reads_all      0\n",
       "pos            0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl.isnull().sum().sort_values(ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A_fraction    18\n",
       "A             18\n",
       "G_fraction    16\n",
       "G             16\n",
       "C_fraction    15\n",
       "C             15\n",
       "T_fraction    14\n",
       "T             14\n",
       "insertions     0\n",
       "deletions      0\n",
       "mismatches     0\n",
       "matches        0\n",
       "reads_all      0\n",
       "pos            0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl.isna().sum().sort_values(ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>reads_all</th>\n",
       "      <th>matches</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>deletions</th>\n",
       "      <th>insertions</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>A_fraction</th>\n",
       "      <th>T_fraction</th>\n",
       "      <th>G_fraction</th>\n",
       "      <th>C_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.867511</td>\n",
       "      <td>0.313912</td>\n",
       "      <td>-0.275308</td>\n",
       "      <td>-0.245276</td>\n",
       "      <td>0.083401</td>\n",
       "      <td>-0.210161</td>\n",
       "      <td>0.184132</td>\n",
       "      <td>-0.172507</td>\n",
       "      <td>-0.319078</td>\n",
       "      <td>-0.210598</td>\n",
       "      <td>-0.174640</td>\n",
       "      <td>-0.319882</td>\n",
       "      <td>0.183674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads_all</th>\n",
       "      <td>0.867511</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.324307</td>\n",
       "      <td>-0.293385</td>\n",
       "      <td>-0.241827</td>\n",
       "      <td>0.130187</td>\n",
       "      <td>-0.144768</td>\n",
       "      <td>0.042765</td>\n",
       "      <td>-0.101148</td>\n",
       "      <td>-0.390610</td>\n",
       "      <td>-0.145137</td>\n",
       "      <td>-0.102624</td>\n",
       "      <td>-0.391164</td>\n",
       "      <td>0.040941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>matches</th>\n",
       "      <td>0.313912</td>\n",
       "      <td>0.324307</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.888921</td>\n",
       "      <td>-0.777581</td>\n",
       "      <td>-0.225610</td>\n",
       "      <td>-0.581108</td>\n",
       "      <td>-0.371192</td>\n",
       "      <td>-0.478788</td>\n",
       "      <td>-0.738893</td>\n",
       "      <td>-0.581128</td>\n",
       "      <td>-0.478866</td>\n",
       "      <td>-0.739024</td>\n",
       "      <td>-0.369283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mismatches</th>\n",
       "      <td>-0.275308</td>\n",
       "      <td>-0.293385</td>\n",
       "      <td>-0.888921</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.403204</td>\n",
       "      <td>0.111158</td>\n",
       "      <td>0.856523</td>\n",
       "      <td>0.324956</td>\n",
       "      <td>0.318009</td>\n",
       "      <td>0.781218</td>\n",
       "      <td>0.856547</td>\n",
       "      <td>0.318103</td>\n",
       "      <td>0.781095</td>\n",
       "      <td>0.324238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deletions</th>\n",
       "      <td>-0.245276</td>\n",
       "      <td>-0.241827</td>\n",
       "      <td>-0.777581</td>\n",
       "      <td>0.403204</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.299234</td>\n",
       "      <td>-0.069381</td>\n",
       "      <td>0.287885</td>\n",
       "      <td>0.528187</td>\n",
       "      <td>0.482378</td>\n",
       "      <td>-0.069377</td>\n",
       "      <td>0.528206</td>\n",
       "      <td>0.482804</td>\n",
       "      <td>0.285142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insertions</th>\n",
       "      <td>0.083401</td>\n",
       "      <td>0.130187</td>\n",
       "      <td>-0.225610</td>\n",
       "      <td>0.111158</td>\n",
       "      <td>0.299234</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.010117</td>\n",
       "      <td>0.286595</td>\n",
       "      <td>0.001270</td>\n",
       "      <td>0.094896</td>\n",
       "      <td>-0.010385</td>\n",
       "      <td>0.001462</td>\n",
       "      <td>0.095061</td>\n",
       "      <td>0.285734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>-0.210161</td>\n",
       "      <td>-0.144768</td>\n",
       "      <td>-0.581108</td>\n",
       "      <td>0.856523</td>\n",
       "      <td>-0.069381</td>\n",
       "      <td>-0.010117</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.129633</td>\n",
       "      <td>-0.121939</td>\n",
       "      <td>0.265298</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>-0.121539</td>\n",
       "      <td>0.264570</td>\n",
       "      <td>-0.126436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.184132</td>\n",
       "      <td>0.042765</td>\n",
       "      <td>-0.371192</td>\n",
       "      <td>0.324956</td>\n",
       "      <td>0.287885</td>\n",
       "      <td>0.286595</td>\n",
       "      <td>-0.129633</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.174355</td>\n",
       "      <td>0.158787</td>\n",
       "      <td>-0.129703</td>\n",
       "      <td>0.174420</td>\n",
       "      <td>0.158192</td>\n",
       "      <td>0.999958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>-0.172507</td>\n",
       "      <td>-0.101148</td>\n",
       "      <td>-0.478788</td>\n",
       "      <td>0.318009</td>\n",
       "      <td>0.528187</td>\n",
       "      <td>0.001270</td>\n",
       "      <td>-0.121939</td>\n",
       "      <td>0.174355</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.096304</td>\n",
       "      <td>-0.121835</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>0.097013</td>\n",
       "      <td>0.171335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>-0.319078</td>\n",
       "      <td>-0.390610</td>\n",
       "      <td>-0.738893</td>\n",
       "      <td>0.781218</td>\n",
       "      <td>0.482378</td>\n",
       "      <td>0.094896</td>\n",
       "      <td>0.265298</td>\n",
       "      <td>0.158787</td>\n",
       "      <td>0.096304</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.265681</td>\n",
       "      <td>0.096495</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>0.156450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_fraction</th>\n",
       "      <td>-0.210598</td>\n",
       "      <td>-0.145137</td>\n",
       "      <td>-0.581128</td>\n",
       "      <td>0.856547</td>\n",
       "      <td>-0.069377</td>\n",
       "      <td>-0.010385</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>-0.129703</td>\n",
       "      <td>-0.121835</td>\n",
       "      <td>0.265681</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.121436</td>\n",
       "      <td>0.264947</td>\n",
       "      <td>-0.126499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_fraction</th>\n",
       "      <td>-0.174640</td>\n",
       "      <td>-0.102624</td>\n",
       "      <td>-0.478866</td>\n",
       "      <td>0.318103</td>\n",
       "      <td>0.528206</td>\n",
       "      <td>0.001462</td>\n",
       "      <td>-0.121539</td>\n",
       "      <td>0.174420</td>\n",
       "      <td>0.999984</td>\n",
       "      <td>0.096495</td>\n",
       "      <td>-0.121436</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.097204</td>\n",
       "      <td>0.171400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G_fraction</th>\n",
       "      <td>-0.319882</td>\n",
       "      <td>-0.391164</td>\n",
       "      <td>-0.739024</td>\n",
       "      <td>0.781095</td>\n",
       "      <td>0.482804</td>\n",
       "      <td>0.095061</td>\n",
       "      <td>0.264570</td>\n",
       "      <td>0.158192</td>\n",
       "      <td>0.097013</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>0.264947</td>\n",
       "      <td>0.097204</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.155852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C_fraction</th>\n",
       "      <td>0.183674</td>\n",
       "      <td>0.040941</td>\n",
       "      <td>-0.369283</td>\n",
       "      <td>0.324238</td>\n",
       "      <td>0.285142</td>\n",
       "      <td>0.285734</td>\n",
       "      <td>-0.126436</td>\n",
       "      <td>0.999958</td>\n",
       "      <td>0.171335</td>\n",
       "      <td>0.156450</td>\n",
       "      <td>-0.126499</td>\n",
       "      <td>0.171400</td>\n",
       "      <td>0.155852</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 pos  reads_all   matches  mismatches  deletions  insertions  \\\n",
       "pos         1.000000   0.867511  0.313912   -0.275308  -0.245276    0.083401   \n",
       "reads_all   0.867511   1.000000  0.324307   -0.293385  -0.241827    0.130187   \n",
       "matches     0.313912   0.324307  1.000000   -0.888921  -0.777581   -0.225610   \n",
       "mismatches -0.275308  -0.293385 -0.888921    1.000000   0.403204    0.111158   \n",
       "deletions  -0.245276  -0.241827 -0.777581    0.403204   1.000000    0.299234   \n",
       "insertions  0.083401   0.130187 -0.225610    0.111158   0.299234    1.000000   \n",
       "A          -0.210161  -0.144768 -0.581108    0.856523  -0.069381   -0.010117   \n",
       "C           0.184132   0.042765 -0.371192    0.324956   0.287885    0.286595   \n",
       "T          -0.172507  -0.101148 -0.478788    0.318009   0.528187    0.001270   \n",
       "G          -0.319078  -0.390610 -0.738893    0.781218   0.482378    0.094896   \n",
       "A_fraction -0.210598  -0.145137 -0.581128    0.856547  -0.069377   -0.010385   \n",
       "T_fraction -0.174640  -0.102624 -0.478866    0.318103   0.528206    0.001462   \n",
       "G_fraction -0.319882  -0.391164 -0.739024    0.781095   0.482804    0.095061   \n",
       "C_fraction  0.183674   0.040941 -0.369283    0.324238   0.285142    0.285734   \n",
       "\n",
       "                   A         C         T         G  A_fraction  T_fraction  \\\n",
       "pos        -0.210161  0.184132 -0.172507 -0.319078   -0.210598   -0.174640   \n",
       "reads_all  -0.144768  0.042765 -0.101148 -0.390610   -0.145137   -0.102624   \n",
       "matches    -0.581108 -0.371192 -0.478788 -0.738893   -0.581128   -0.478866   \n",
       "mismatches  0.856523  0.324956  0.318009  0.781218    0.856547    0.318103   \n",
       "deletions  -0.069381  0.287885  0.528187  0.482378   -0.069377    0.528206   \n",
       "insertions -0.010117  0.286595  0.001270  0.094896   -0.010385    0.001462   \n",
       "A           1.000000 -0.129633 -0.121939  0.265298    0.999998   -0.121539   \n",
       "C          -0.129633  1.000000  0.174355  0.158787   -0.129703    0.174420   \n",
       "T          -0.121939  0.174355  1.000000  0.096304   -0.121835    0.999984   \n",
       "G           0.265298  0.158787  0.096304  1.000000    0.265681    0.096495   \n",
       "A_fraction  0.999998 -0.129703 -0.121835  0.265681    1.000000   -0.121436   \n",
       "T_fraction -0.121539  0.174420  0.999984  0.096495   -0.121436    1.000000   \n",
       "G_fraction  0.264570  0.158192  0.097013  0.999995    0.264947    0.097204   \n",
       "C_fraction -0.126436  0.999958  0.171335  0.156450   -0.126499    0.171400   \n",
       "\n",
       "            G_fraction  C_fraction  \n",
       "pos          -0.319882    0.183674  \n",
       "reads_all    -0.391164    0.040941  \n",
       "matches      -0.739024   -0.369283  \n",
       "mismatches    0.781095    0.324238  \n",
       "deletions     0.482804    0.285142  \n",
       "insertions    0.095061    0.285734  \n",
       "A             0.264570   -0.126436  \n",
       "C             0.158192    0.999958  \n",
       "T             0.097013    0.171335  \n",
       "G             0.999995    0.156450  \n",
       "A_fraction    0.264947   -0.126499  \n",
       "T_fraction    0.097204    0.171400  \n",
       "G_fraction    1.000000    0.155852  \n",
       "C_fraction    0.155852    1.000000  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nucl.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-27-7359c8236f70>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m h_labels = [x.replace('_', ' ').title() for x in \n\u001b[1;32m      2\u001b[0m             list(nucl)]\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mfig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m6\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheatmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnucl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcorr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mannot\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mxticklabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mh_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0myticklabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mh_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcubehelix_palette\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mas_cmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "h_labels = [x.replace('_', ' ').title() for x in \n",
    "            list(nucl)]\n",
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "_ = sns.heatmap(nucl.corr(), annot=True, xticklabels=h_labels, yticklabels=h_labels, cmap=sns.cubehelix_palette(as_cmap=True), ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Additional task\n",
    "# Make functions\n",
    "def read_gff(path):\n",
    "    gff = pd.read_csv(path, sep='\\t', comment='#', skiprows=1)\n",
    "    gff.columns = ['chromosome', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    return(gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>68981</td>\n",
       "      <td>72061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=23S_rRNA;product=23S ribosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>469298</td>\n",
       "      <td>470826</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=16S_rRNA;product=16S ribosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>471166</td>\n",
       "      <td>474246</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=23S_rRNA;product=23S ribosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3629599</td>\n",
       "      <td>3632680</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=23S_rRNA;product=23S ribosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3633020</td>\n",
       "      <td>3634549</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=16S_rRNA;product=16S ribosomal RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome       source  type    start      end  score strand phase  \\\n",
       "0  Reference_10  barrnap:0.9  rRNA    68981    72061    0.0      +     .   \n",
       "1  Reference_10  barrnap:0.9  rRNA   469298   470826    0.0      +     .   \n",
       "2  Reference_10  barrnap:0.9  rRNA   471166   474246    0.0      +     .   \n",
       "3  Reference_10  barrnap:0.9  rRNA  3629599  3632680    0.0      -     .   \n",
       "4  Reference_10  barrnap:0.9  rRNA  3633020  3634549    0.0      -     .   \n",
       "\n",
       "                                attributes  \n",
       "0  Name=23S_rRNA;product=23S ribosomal RNA  \n",
       "1  Name=16S_rRNA;product=16S ribosomal RNA  \n",
       "2  Name=23S_rRNA;product=23S ribosomal RNA  \n",
       "3  Name=23S_rRNA;product=23S ribosomal RNA  \n",
       "4  Name=16S_rRNA;product=16S ribosomal RNA  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = read_gff('/Users/Sasha/Downloads/rrna_annotation.gff')\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bed6(path):\n",
    "    bed = pd.read_csv(path, sep='\\t', comment='#', skiprows=1)\n",
    "    bed.columns = ['chromosome', 'start', 'end', 'name', 'score', 'strand']\n",
    "    return bed\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Reference_1</td>\n",
       "      <td>202941</td>\n",
       "      <td>203377</td>\n",
       "      <td>NODE_5666_length_436_cov_418.653543</td>\n",
       "      <td>16</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Reference_1</td>\n",
       "      <td>203322</td>\n",
       "      <td>785222</td>\n",
       "      <td>NODE_25_length_581906_cov_146.024228</td>\n",
       "      <td>60</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Reference_1</td>\n",
       "      <td>785169</td>\n",
       "      <td>1763981</td>\n",
       "      <td>NODE_11_length_978812_cov_132.517515</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Reference_1</td>\n",
       "      <td>1763926</td>\n",
       "      <td>1764363</td>\n",
       "      <td>NODE_5666_length_436_cov_418.653543</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Reference_1</td>\n",
       "      <td>1764308</td>\n",
       "      <td>1764418</td>\n",
       "      <td>NODE_19543_length_110_cov_1549.290909</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome    start      end                                   name  \\\n",
       "0  Reference_1   202941   203377    NODE_5666_length_436_cov_418.653543   \n",
       "1  Reference_1   203322   785222   NODE_25_length_581906_cov_146.024228   \n",
       "2  Reference_1   785169  1763981   NODE_11_length_978812_cov_132.517515   \n",
       "3  Reference_1  1763926  1764363    NODE_5666_length_436_cov_418.653543   \n",
       "4  Reference_1  1764308  1764418  NODE_19543_length_110_cov_1549.290909   \n",
       "\n",
       "   score strand  \n",
       "0     16      +  \n",
       "1     60      +  \n",
       "2     60      -  \n",
       "3      0      -  \n",
       "4      0      +  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed = read_bed6('/Users/Sasha/Downloads/alignment.bed')\n",
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>68981</td>\n",
       "      <td>72061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>23S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>469298</td>\n",
       "      <td>470826</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>16S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>471166</td>\n",
       "      <td>474246</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>23S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3629599</td>\n",
       "      <td>3632680</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>23S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Reference_10</td>\n",
       "      <td>barrnap:0.9</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3633020</td>\n",
       "      <td>3634549</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>16S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome       source  type    start      end  score strand phase  \\\n",
       "0  Reference_10  barrnap:0.9  rRNA    68981    72061    0.0      +     .   \n",
       "1  Reference_10  barrnap:0.9  rRNA   469298   470826    0.0      +     .   \n",
       "2  Reference_10  barrnap:0.9  rRNA   471166   474246    0.0      +     .   \n",
       "3  Reference_10  barrnap:0.9  rRNA  3629599  3632680    0.0      -     .   \n",
       "4  Reference_10  barrnap:0.9  rRNA  3633020  3634549    0.0      -     .   \n",
       "\n",
       "  attributes  \n",
       "0        23S  \n",
       "1        16S  \n",
       "2        23S  \n",
       "3        23S  \n",
       "4        16S  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract simple names of rRNA\n",
    "import re\n",
    "rna_info = read_gff('/Users/Sasha/Downloads/rrna_annotation.gff')\n",
    "rna_info_convenient = rna_info\n",
    "rna_info_convenient['attributes']=rna_info['attributes'].str.extract('(Name.*rRNA)')\n",
    "rna_info_convenient['attributes'] = rna_info_convenient['attributes'].str.replace('Name=', '')\n",
    "rna_info_convenient['attributes'] = rna_info_convenient['attributes'].str.replace('_rRNA', '')\n",
    "rna_info_convenient.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome</th>\n",
       "      <th>attributes</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Reference_1</th>\n",
       "      <th>16S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Reference_10</th>\n",
       "      <th>16S</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23S</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Reference_8</th>\n",
       "      <th>16S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Reference_9</th>\n",
       "      <th>16S</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23S</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5S</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         count\n",
       "chromosome   attributes       \n",
       "Reference_1  16S             2\n",
       "             23S             2\n",
       "             5S              2\n",
       "Reference_10 16S             4\n",
       "             23S             5\n",
       "...                        ...\n",
       "Reference_8  16S             2\n",
       "             23S             2\n",
       "Reference_9  16S             6\n",
       "             23S             6\n",
       "             5S              6\n",
       "\n",
       "[76 rows x 1 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count distinct rRNA types per chromosome\n",
    "rna_dist = rna_info_convenient.filter(items=['chromosome','attributes'])\n",
    "rna_dist['count'] = 1\n",
    "rna_dist.head()\n",
    "rna_dist.groupby(['chromosome','attributes']).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6020</th>\n",
       "      <td>Reference_20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>91528</td>\n",
       "      <td>60.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_149_length_138283_cov_350.133699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2942</th>\n",
       "      <td>Reference_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>364810</td>\n",
       "      <td>60.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_36_length_466911_cov_271.254873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6779</th>\n",
       "      <td>Reference_23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>160865</td>\n",
       "      <td>60.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_80_length_244836_cov_368.038496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>Reference_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>128866</td>\n",
       "      <td>60.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_158_length_133586_cov_762.631164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2942</th>\n",
       "      <td>Reference_8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>364810</td>\n",
       "      <td>60.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NODE_36_length_466911_cov_271.254873</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome source type  start     end  score strand phase attributes  \\\n",
       "6020  Reference_20    NaN  NaN      0   91528   60.0      +   NaN        NaN   \n",
       "2942   Reference_8    NaN  NaN      0  364810   60.0      +   NaN        NaN   \n",
       "6779  Reference_23    NaN  NaN      0  160865   60.0      -   NaN        NaN   \n",
       "914    Reference_3    NaN  NaN      0  128866   60.0      -   NaN        NaN   \n",
       "2942   Reference_8    NaN  NaN      0  364810   60.0      +   NaN        NaN   \n",
       "\n",
       "                                       name  \n",
       "6020  NODE_149_length_138283_cov_350.133699  \n",
       "2942   NODE_36_length_466911_cov_271.254873  \n",
       "6779   NODE_80_length_244836_cov_368.038496  \n",
       "914   NODE_158_length_133586_cov_762.631164  \n",
       "2942   NODE_36_length_466911_cov_271.254873  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Intersect\n",
    "merged = gff.append(bed)\n",
    "merged = merged.sort_values(by='start')\n",
    "merged.head()\n",
    "annot_ind_list = merged.index[merged['type'] != 'NaN'].tolist()\n",
    "annot_read_start = [x+1 for x in annot_ind_list]\n",
    "annot_read_end = [x-1 for x in annot_ind_list]\n",
    "save = annot_read_start + annot_read_end + annot_ind_list\n",
    "merged1 = merged.iloc[save]\n",
    "merged1.sort_values(by='start').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
